### PR TITLE
storage: rework storage drivers

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by brauner** [comment]
Marking as incomplete since the other backends need to be adapted as well.

@overlayfs, if you want you can grab these commits and test at least for btrfs. Fixes the issues for me here. Sorry that it took longer. There's a bunch of stuff I need to take care of. But don't worry we won't release  `2.1` until we're sure that things work fine!

**Comment by overlayfs** [comment]
Hmmm, maybe I grabbed the wrong code?  I used the _download zip_ option from [this link](https://github.com/lxc/lxc/tree/465c2f8091b8f35d2b7c33c8717e0c3f57d6bf43) .

`lxc-copy -B btrfs -n trusty -N debug`
fails with error message,

> lxc-copy: trusty: bdev/bdev.c: bdev_copy: 440 Error restoring /home/ubuntu/.local/share/lxc/trusty/rootfs to /home/ubuntu/.local/share/lxc/debug/rootfs
> lxc-copy: trusty: lxccontainer.c: copy_storage: 3070 Error copying storage.
> clone failed

[debug.log.txt](https://github.com/lxc/lxc/files/1150535/debug.log.txt)

**Comment by brauner** [comment]
@overlayfs, seems like it. What distro are you on?

**Comment by overlayfs** [comment]
ubuntu-14.04 with lxc-daily ppa, linux-generic-lts-xenial kernel and btrfs.

**Comment by overlayfs** [comment]
**trusty/config**
lxc.include = /usr/share/lxc/config/ubuntu.common.conf
lxc.include = /usr/share/lxc/config/ubuntu.userns.conf
lxc.arch = x86_64
lxc.include = /etc/lxc/default.conf
lxc.id_map = u 0 100000 65536
lxc.id_map = g 0 100000 65536
lxc.rootfs.path = btrfs:/home/ubuntu/.local/share/lxc/trusty/rootfs
lxc.utsname = trusty

**Comment by brauner** [comment]
@overlayfs, cool so what I usually do is:
```
./autogen.sh && ./configure --build=x86_64-linux-gnu --prefix=/usr --includedir=\${prefix}/include --mandir=\${prefix}/share/man --infodir=\${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-silent-rules --libdir=\${prefix}/lib/x86_64-linux-gnu --libexecdir=\${prefix}/lib/x86_64-linux-gnu --disable-maintainer-mode --disable-dependency-tracking --libdir=\${prefix}/lib/x86_64-linux-gnu --libexecdir=\${prefix}/lib/x86_64-linux-gnu --with-rootfs-path=\${prefix}/lib/x86_64-linux-gnu/lxc "CFLAGS=-g -O2 -fdebug-prefix-map=/«PKGBUILDDIR»=. -fstack-protector-strong -Wformat -Werror=format-security" "CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2" "CXXFLAGS=-g -O2 -fdebug-prefix-map=/«PKGBUILDDIR»=. -fstack-protector-strong -Wformat -Werror=format-security" "FCFLAGS=-g -O2 -fdebug-prefix-map=/«PKGBUILDDIR»=. -fstack-protector-strong" "FFLAGS=-g -O2 -fdebug-prefix-map=/«PKGBUILDDIR»=. -fstack-protector-strong" "GCJFLAGS=-g -O2 -fdebug-prefix-map=/«PKGBUILDDIR»=. -fstack-protector-strong" "LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro" "OBJCFLAGS=-g -O2 -fdebug-prefix-map=/«PKGBUILDDIR»=. -fstack-protector-strong -Wformat -Werror=format-security" "OBJCXXFLAGS=-g -O2 -fdebug-prefix-map=/«PKGBUILDDIR»=. -fstack-protector-strong -Wformat -Werror=format-security" --enable-doc --disable-rpath --enable-apparmor --enable-selinux --enable-tests --with-distro=ubuntu && make -j4
```

and then in the shell I want to test in do:

    export LD_LIBRARY_PATH=/the/path/to/lxc/src/lxc/.libs

and then do the tests.

But you could also do:

    apt source lxc
    apt build-dep lxc
    cp -r /path/to/lxc/* /path/to/package/dir/

then change to the package dir and do

    debuild -b -uc -us

then

    sudo dpkg -i ../*.deb

**Comment by overlayfs** [comment]
I'll try rebuilding it tomorrow.  I made use of github/lxc/lxc-pkg-ubuntu, and I think I inadvertently used the ubuntu-16.04 version rather than the ubuntu-14.04 version.

**Comment by brauner** [comment]
jenkins: test this please

**Comment by overlayfs** [comment]
The following commands succeed:
```
lxc-copy -s -B btrfs -n trusty -N debug
lxc-copy -s -B btrfs -n trusty -N debug -p /tmp
```
The following command fails (see debug.log.txt above for the error message):
`lxc-copy -B btrfs -n trusty -N debug`

**Comment by brauner** [comment]
@overlayfs, thanks for testing! :)

**Comment by brauner** [comment]
@overlayfs, if you want you can test again, I redid parts of the btrfs backend.

**Comment by overlayfs** [comment]
TLDR; I tested lxc-copy at commit 99f6a6e and everything seems to work (-:

Long version:

**Using the new syntax:** 
**lxc.rootfs.path = btrfs:/home/ubuntu/.local/share/lxc/trusty/rootfs**
- `lxc-copy -n trusty -N debug -p newpath`
creates a btrfs-snapshot-based container.

**Using the old (obsolete) syntax:** 
**lxc.rootfs = /home/ubuntu/.local/share/lxc/trusty/rootfs**
- `lxc-copy -n trusty -N debug -p newpath`
`lxc-copy -s -n trusty -N debug -p newpath`
creates an overlayfs-based container.
- `lxc-copy -B btrfs -n trusty -N debug -p newpath`
`lxc-copy -s -B btrfs  -n trusty -N debug -p newpath`
produces an error message,
> lxc-copy: debug: bdev/lxcbtrfs.c: btrfs_clonepaths: 385 btrfs snapshot from dir backing store is not supported
> lxc-copy: debug: bdev/bdev.c: bdev_copy: 432 Failed getting pathnames for clone of "/home/ubuntu/.local/share/lxc/debug/rootfs"

This error-message seems reasonable since the default backing-store is dir.  But it does constitute a change in behaviour, since this syntax used to work.  So it might be worth printing a longer error message which helps the user to resolve the problem. (Eg perhaps mention that lxc.rootfs should be replaced by lxc.rootfs.path).

**Comment by brauner** [comment]
@overlayfs, thanks for testing! Ok, so 
```
lxc-copy -s -B btrfs -n trusty -N debug -p newpath
```
shouldn't work if `-p newpath` specifies a non-`btrfs` path and `-B btrfs` means requesting a `btrfs` snapshot.

**Comment by brauner** [comment]
Reworked backends:
- [x] `btrfs`
- [x] `lvm`

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

File: src/lxc/utils.c [hunk 1]
```diff
@@ -728,47 +728,49 @@ char **lxc_normalize_path(const char *path)
...
+
 			p = strdup("/");
 			if (!p)
-				goto out;
-			free(*path);
-			*path = p;
-			ret = true;
-			goto out;
+				return NULL;
+
```
Comment (hallyn) [inline src/lxc/utils.c]: Just 'return p' here?  :-)

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
@hallyn, updated.

**Comment by hallyn** [comment]
Overall LGTM, thanks.

**Comment by brauner** [comment]
jenkins: test this please
